src/ray/core_worker/transport/direct_actor_task_submitter.cc

```diff
@@ -339,26 +343,27 @@ void CoreWorkerDirectActorTaskSubmitter::DisconnectActor(
 }
 
 void CoreWorkerDirectActorTaskSubmitter::CheckTimeoutTasks() {
-  std::vector<TaskSpecification> task_specs;
+  std::vector<std::pair<TaskSpecification, Status>> task_specs_and_status;
   {
     absl::MutexLock lock(&mu_);
     for (auto &queue_pair : client_queues_) {
       auto &queue = queue_pair.second;
       auto deque_itr = queue.wait_for_death_info_tasks.begin();
       while (deque_itr != queue.wait_for_death_info_tasks.end() &&
              /*timeout timestamp*/ deque_itr->first < current_time_ms()) {
-        auto &task_spec = deque_itr->second;
-        task_specs.push_back(task_spec);
+        auto &task_spec_status_pair = deque_itr->second;
+        task_specs_and_status.push_back(task_spec_status_pair);
         deque_itr = queue.wait_for_death_info_tasks.erase(deque_itr);
       }
     }
   }
 
   // Do not hold mu_, because FailPendingTask may call python from cpp,
   // and may cause deadlock with SubmitActorTask thread when aquire GIL.
-  for (auto &task_spec : task_specs) {
-    GetTaskFinisherWithoutMu().FailPendingTask(task_spec.TaskId(),
-                                               rpc::ErrorType::ACTOR_DIED);
+  for (auto &task_spec_status_pair : task_specs_and_status) {
+    GetTaskFinisherWithoutMu().FailPendingTask(task_spec_status_pair.first.TaskId(),
+                                               rpc::ErrorType::ACTOR_DIED,
+                                               &task_spec_status_pair.second);
```

Comment:
what's the impact of this change btw?

Reply:
The impact of this change is such that actor methods which fail with intentional system exit would not be marked as failure, including: 
1. exit_actor
2. exit(0)

(not the `max_calls` case since that applies to task only.)

Reply:
oh I am specifically asking for passing "status". So it is necessary to detect the status code?

Reply:
one concern is when I see this API FailPendingTask, we are picking the status message not the message from the error_info.

Reply:
For example this part,

```
      for (auto &net_err_task_pair : wait_for_death_info_tasks) {
        RAY_UNUSED(GetTaskFinisherWithoutMu().FailPendingTask(
            net_err_task_pair.second.first.TaskId(),
            error_type,
            /* status */ &net_err_task_pair.second.second,
            &error_info));
```

I think we should prioritize error_info error message over status?

Reply:
```
    SetTaskStatus(
        it->second,
        rpc::TaskStatus::FAILED,
        (ray_error_info == nullptr
             ? gcs::GetRayErrorInfo(error_type, (status ? status->ToString() : ""))
             : *ray_error_info));
```

Reply:
> So it is necessary to detect the status code?

Yeah, we need this (both error info, and status) due to a combo of legacy API + multiple errors places. 

> one concern is when I see this API FailPendingTask, we are picking the status message not the message from the error_info.

We are using the error message from `error_info` rather than the status. Do you think we should pick the other? An example scenario, when `exit_actor` happens: 
- status: will contain info of worker exit and intentional system exit
- ray_error_info: will contain the actor_died info, and also the worker exit message. 

I think it doesn't make a huge difference between choosing one or the other. but i agree this API is a bit confusing...

Reply:
OH we should prioritize error_info. I think I misread the code.